### PR TITLE
feat(model): add `Model.applyVirtuals()` to apply virtuals to a POJO

### LINKED ISSUE #14818: Model.applyVirtuals(): apply schema virtuals to a POJO, similar to Model.applyDefaults()
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the feature has not already been requested

### 🚀 Feature Proposal

Supporting virtual fields in aggregate will be good feature.

I know aggregate is server-side feature and virtual fields is client-side feature and aggregation results can change shape of document.
However, some aggregate results just extends document data and it is possible to add virtual fields to aggregate results, just I mentioned in motivation.
I think we can add `hasVirtualFields` to aggregate options which default is false and add virtual fields to aggregate results.

### Motivation

I am using aggregate function for pagination. 
Since aggregate doesn't support virtual fields, I need to hydrate each aggregation result and add virtual fields to each aggregation result manually.
It will be good to do in mongoose itself.

### Example

```
UserModel.aggregate([
  // add pipelines
], {
  hasVirtualFields: true
});
```

So, we can add virtual fields like `fullName` to aggregate results.

### PR DESCRIPTION
**Pull request by vkarpov15**
Fix #14818

**Summary**

Logic largely copied from https://github.com/mongoosejs/mongoose-lean-virtuals/blob/master/index.js to apply a Model's virtuals to a POJO.

**Examples**

### CODE REVIEW

File: lib/model.js [hunk 1]
```diff
@@ -3503,6 +3507,40 @@ Model.applyDefaults = function applyDefaults(doc) {
...
+ *     const userSchema = new Schema({ name: String });
+ *     userSchema.virtual('upper').get(function() { return this.name.toUpperCase(); });
+ *     const User = mongoose.model('User', userSchema);
+ *
+ *     const obj = { name: 'John' };
+ *     User.applyVirtuals(obj);
+ *     obj.name; // 'John'
+ *     obj.upper; // 'JOHN', Mongoose applied the return value of the virtual to the given object
+ *
+ * @param {Object} obj object or document to apply virtuals on
```
Comment (hasezoey) [inline lib/model.js:3524]:
```suggestion
 * @param {Object} doc object or document to apply virtuals on
```

consistency with the parameter name

File: lib/model.js [hunk 2]
```diff
@@ -3503,6 +3507,40 @@ Model.applyDefaults = function applyDefaults(doc) {
...
+ * @param {Array<string>} [virtualsToApply] optional whitelist of virtuals to apply
+ * @returns {Object} obj
```
Comment (hasezoey) [inline lib/model.js:3526]:
```suggestion
 * @returns {Object} doc
```

File: test/model.test.js [hunk 3]
```diff
@@ -7781,6 +7781,208 @@ describe('Model', function() {
...
+          first: 'Bill',
+          last: 'Gates'
+        }
+      }, ['name.lastLower']);
+      assert.strictEqual(res.name.first, 'Bill');
+      assert.strictEqual(res.name.last, 'Gates');
+      assert.strictEqual(res.fullName, undefined);
+      assert.strictEqual(res.name.firstUpper, undefined);
+      assert.strictEqual(res.name.lastLower, 'gates');
+    });
```
Comment (hasezoey) [inline test/model.test.js:7949]: isnt this literally a copy of the test before it?

File: lib/helpers/document/applyVirtuals.js [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+      continue;
+    }
+    const virtualType = schema.virtuals[virtual];
+    const sp = Array.isArray(virtual)
+      ? virtual :
+      virtual.indexOf('.') === -1
+        ? [virtual]
+        : virtual.split('.');
+    let cur = doc;
+    for (let j = 0; j < sp.length - 1; ++j) {
```
Comment (hasezoey) [inline lib/helpers/document/applyVirtuals.js:123]: just to confirm, this is meant to apply to all but the last `sp`? (to my knowledge `<` is exclusive, like the other for loops are using it)

**Review by hasezoey — COMMENTED** [review]
mostly looks good to me, some minor style suggestions, aside from that:
- typescript types for the new `applyVirtuals` are not added yet
- there is a distinct lack of documentation in the `applyVirtuals` file

File: test/model.test.js (hunk 3 above)
Comment (vkarpov15) [reply test/model.test.js:7906]: Good catch, sorry about that. Fixed.

File: lib/helpers/document/applyVirtuals.js (hunk 4 above)
Comment (vkarpov15) [reply lib/helpers/document/applyVirtuals.js:123]: Yes, this loop is meant to get the last nested object in the path so we can set `cur[sp.length - 1]`.

**Comment by vkarpov15** [comment]
@hasezoey thanks for the feedback. I added docs and addressed issues :+1:

File: lib/model.js (hunk 2 above)
Comment (vkarpov15) [reply lib/model.js:3526]: I changed consistently to `obj` because `doc` implies the parameter should be a document, which isn't the case.

**Review by hasezoey — COMMENTED** [review]
Looks good to me, typescript types for the new `applyVirtuals` on `Model` are still missing though

**Review by hasezoey — COMMENTED** [review]
Looks good to me

**Pull request merged**